server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -82,103 +82,112 @@ class ObjectService(
+                    for {
+                      cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath))
+                      status <- cachedGeneration match {
+                        case Some(cachedGen) => if(cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
+                        case None => IO.pure(SyncStatus.OutOfSync)
+                      }
+                    } yield status
+                  }
+                  currentTime <- timer.clock.realTime(TimeUnit.MILLISECONDS)
+                  lastLock <- expiresAt.flatTraverse[IO, WorkbenchEmail] { ea =>
+                    if (currentTime > ea.toEpochMilli)
+                      IO.pure(none[WorkbenchEmail])
+                    else
+                      IO.pure(lastLockedBy)
+                  }
+                } yield MetadataResponse.EditMode(syncStatus, lastLock, generation, sl)
+            }
+          } yield result
         }
-      } yield result
-    }
+      }
+    } yield res
 
   def safeDelocalize(req: SafeDelocalize): IO[Unit] =
     for {
       traceId <- IO(TraceId(randomUUID()))
-      storageLinks <- storageLinksCache.get
-      baseDirectory <- IO.fromEither(getLocalBaseDirectory(req.localObjectPath).leftMap(s => BadRequestException(s)))
-      storageLink = storageLinks.find(x => x._1.path == baseDirectory)
+      storageLink <- findStorageLink(req.localObjectPath)
       _ <- storageLink.fold[IO[Unit]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${req.localObjectPath}"))) { pair =>
+        val isSafeMode = pair._1
         val sl = pair._2
-        for {
-          previousMeta <- metadataCache.get.map(_.get(req.localObjectPath))
-          fullBlobName <- IO.fromEither(getFullBlobName(req.localObjectPath, sl.cloudStorageDirectory.blobPath).leftMap(s => BadRequestException(s)))
-          latestMetadata <- retrieveGcsMetadata(sl.cloudStorageDirectory.bucketName, fullBlobName, traceId)
-          gsPath = GsPath(sl.cloudStorageDirectory.bucketName, fullBlobName)
-          _ <- (previousMeta, latestMetadata) match {
-            case (Some(prev), Some(now)) =>
-              if(prev.crc32c == now.crc32c)
-                delocalize(req.localObjectPath, gsPath, traceId, prev.generation)
-              else
-                IO.raiseError(FileOutOfSyncException("file is updated in google storage by someone else"))
-            case (_, None) => // if file is deleted in remote, we'll just delocalize the file
-              delocalize(req.localObjectPath, gsPath, traceId, 0L)
-            case (None, Some(now)) =>
-              IO.raiseError(new Exception("Can't determine if the file user started off of is up-to-date"))
-          }
-        } yield ()
+        if(isSafeMode)
+          IO.unit
+        else {
+          for {
+            previousMeta <- metadataCache.get.map(_.get(req.localObjectPath))
```

Comment:
Ideally this would populate on a cache miss, rather than assuming no file in GCS.

Reply:
What metadata would you populate the cache with in this case? (We can't just make a google call and get the metadata because it wouldn't be the metadata for the local version. Gcs metadata cache is populated whenever a file is localized)

Reply:
This ties into the other discussion, my comment doesn't make sense if you're planning to precache metadata for every localized file.